Public/Src/Cache/MemoizationStore/Library/Stores/MemoizationStoreConfiguration.cs

```diff
@@ -1,17 +1,41 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using BuildXL.Cache.ContentStore.Distributed.NuCache;
 using BuildXL.Cache.ContentStore.Interfaces.FileSystem;
 using BuildXL.Cache.ContentStore.SQLite;
 using BuildXL.Cache.ContentStore.Stores;
 
 namespace BuildXL.Cache.MemoizationStore.Stores
 {
+    /// <summary>
+    ///     Marker type for children of <see cref="MemoizationStore"/>
+    /// </summary>
+    public class MemoizationStoreConfiguration

```

Comment:
This is kind of ugly, but I could live with it. I think on balance I would prefer to not have these extra marker/wrapper classes, and instead just have your config factory method take an arbitrary object, which it would then test for being either the new RocksDbContentLocationDatabaseConfiguration or the old SqlLiteMemoizationStoreConfiguration and act accordingly. Yes, having an API signature that accepts an arbitrary object and then fails at runtime if it isn't of the right type is ugly, but this new architecture has almost the same problem (since anyone can extend MemoizationStoreConfiguration) and takes a lot more classes/code to do it.

Reply:
It is true that having a marker type is ugly. However, taking an arbitrary object has a series of issues as well:
- There is no semantic information to aid let the reader know what is going on. Receiving a MemoizationStoreConfiguration automatically conveys that somewhere down the road, a MemoizationStore is going to be created.
- Much better extensibility. If in the future we need to add a piece of configuration common to all memoization stores, it is easy this way. With an object, we would need track down every place where a memoization store is instantiated, generate a config object, and progressively introduce the new type.
- The cost is only 1 extra class (namely this one), so it seems to me a small price to pay for avoiding a suboptimal use of time later.

---
In reply to: [306291727](https://github.com/microsoft/BuildXL/pull/647#discussion_r306291727) [](ancestors = 306291727)